test/shared/src/main/scala/zio/test/TestAspect.scala

```diff
@@ -309,17 +309,34 @@ object TestAspect extends TimeoutVariants {
 
   /**
    * An aspect that times out tests using the specified duration.
+   * @param duration maximum test duration
+   * @param interruptDuration after test timeout will wait given duration for successful interruption
    */
-  def timeout(duration: Duration): TestAspect[Nothing, Live[Clock], Nothing, Any, Nothing, Any] =
+  def timeout(
+    duration: Duration,
+    interruptDuration: Duration = 1.second
+  ): TestAspect[Nothing, Live[Clock], Nothing, Any, Nothing, Any] =
     new TestAspect.PerTest[Nothing, Live[Clock], Nothing, Any, Nothing, Any] {
       def perTest[R >: Nothing <: Live[Clock], E >: Nothing <: Any, S >: Nothing <: Any](
         test: ZIO[R, TestFailure[E], TestSuccess[S]]
-      ): ZIO[R, TestFailure[E], TestSuccess[S]] =
-        Live.withLive(test)(_.timeout(duration)).flatMap {
-          case None =>
-            ZIO.fail(TestFailure.Runtime(Cause.die(TestTimeoutException(s"Timeout of ${duration.render} exceeded"))))
-          case Some(v) => ZIO.succeed(v)
+      ): ZIO[R, TestFailure[E], TestSuccess[S]] = {
+        def timeoutFailure =
+          TestFailure.Runtime(Cause.die(TestTimeoutException(s"Timeout of ${duration.render} exceeded.")))
+        def interruptionTimeoutFailure = {
+          val msg =
+            s"Timeout of ${duration.render} exceeded. Couldn't interrupt test within ${interruptDuration.render}, possible resource leak!"
+          TestFailure.Runtime(Cause.die(TestTimeoutException(msg)))
         }
+        for {
+          p <- Promise.make[TestFailure[E], TestSuccess[S]]
+          _ <- test
+                .raceAttempt(Live.live(ZIO.fail(timeoutFailure).delay(duration)))
+                .foldM(p.fail, p.succeed)
+                .fork
+          _      <- (Live.live(ZIO.unit.delay(duration + interruptDuration)) *> p.fail(interruptionTimeoutFailure)).fork
```

Comment:
Don't we want to interrupt this fiber if the first fiber completes successfully? The `Promise` can only be completed once so it won't effect the results but we will have a lot of fibers sitting out there that aren't doing anything.

Reply:
Will do it!